### PR TITLE
FINERACT-2060: re-amortization

### PR DESCRIPTION
**Pull request by taskain7**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 1]
```diff
@@ -182,13 +185,46 @@ public void processLatestTransaction(LoanTransaction loanTransaction, Transactio
...
+        BigDecimal remainingAmount = loanTransaction.getAmount();
+        LocalDate transactionDate = loanTransaction.getTransactionDate();
+        List<LoanRepaymentScheduleInstallment> previousInstallments = installments.stream() //
+                .filter(installment -> !installment.getDueDate().isAfter(transactionDate)) //
+                .toList();
+        List<LoanRepaymentScheduleInstallment> futureInstallments = installments.stream() //
+                .filter(installment -> installment.getDueDate().isAfter(transactionDate)) //
+                .filter(installment -> !installment.isAdditional()) //
+                .toList();
+        previousInstallments.forEach(installment -> {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: I dont really understand this logic.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 2]
```diff
@@ -496,7 +532,15 @@ private void processSingleTransaction(LoanTransaction loanTransaction, MonetaryC
              * For existing transactions, check if the re-payment breakup (principal, interest, fees, penalties) has
              * changed.<br>
              */
-            final LoanTransaction newLoanTransaction = LoanTransaction.copyTransactionProperties(loanTransaction);
+            LoanTransaction newLoanTransaction = LoanTransaction.copyTransactionProperties(loanTransaction);
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:535]: Why do we need this? This PR should not cover the "reverse-replay", no?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/InternalLoanInformationApiResource.java [hunk 3]
```diff
@@ -48,10 +47,8 @@
 import org.apache.fineract.portfolio.loanproduct.data.AdvancedPaymentData;
 import org.apache.fineract.portfolio.loanproduct.mapper.AdvancedPaymentDataMapper;
 import org.springframework.beans.factory.InitializingBean;
-import org.springframework.context.annotation.Profile;
 import org.springframework.stereotype.Component;
 
-@Profile(FineractProfiles.TEST)
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/InternalLoanInformationApiResource.java]: Please undo this.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationServiceImpl.java [hunk 4]
```diff
@@ -117,6 +131,23 @@ private void reverseReAmortizeTransaction(LoanTransaction reAmortizeTransaction,
                 LoanReAmortizationApiConstants.externalIdParameterName);
         reAmortizeTransaction.reverse(reversalExternalId);
         reAmortizeTransaction.manuallyAdjustedOrReversed();
+        reProcessLoanTransactions(reAmortizeTransaction.getLoan());
+    }
+
+    private void reProcessLoanTransactions(Loan loan) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationServiceImpl.java:137]: Why do we need to reprocess the transactions? Reamortization is always the latest transaction, no?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: I think what you need here is:
- collect the outstanding principal in a field
- Remove the outstanding principal from the installment

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 5]
```diff
@@ -496,7 +532,15 @@ private void processSingleTransaction(LoanTransaction loanTransaction, MonetaryC
...
+            LocalDate transactionDate = loanTransaction.getTransactionDate();
+            if (loanTransaction.getTypeOf().equals(REAMORTIZE)) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: I dont think we need this.
I think the original logic should collect the outstanding principal amounts and check whether the new balance is matching with the original transaction amount or not.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/ReAmortizationTest.java [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+import org.apache.fineract.client.models.PostLoanProductsResponse;
+import org.apache.fineract.integrationtests.common.ClientHelper;
+import org.apache.fineract.integrationtests.common.loans.LoanTestLifecycleExtension;
+import org.apache.fineract.portfolio.loanaccount.loanschedule.domain.LoanScheduleProcessingType;
+import org.apache.fineract.portfolio.loanaccount.loanschedule.domain.LoanScheduleType;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(LoanTestLifecycleExtension.class)
+public class ReAmortizationTest extends BaseLoanIntegrationTest {
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/ReAmortizationTest.java]: There's an existing test for reamortization called LoanReAmortizationIntegrationTest

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 1 above)
Comment (taskain7) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: fixed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/InternalLoanInformationApiResource.java (hunk 3 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/InternalLoanInformationApiResource.java]: fixed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 5 above)
Comment (taskain7) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: fixed

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/ReAmortizationTest.java (hunk 6 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/ReAmortizationTest.java]: I moved my test there

**Review by adamsaghy — APPROVED** [review]
LGTM

**Review by galovics — APPROVED** [review]

**Pull request merged**